server/src/network.rs

```diff
@@ -0,0 +1,215 @@
+            let message = ws::Message::Text(String::from_utf8(message.data).unwrap());
+            connection.send(message).log_if_err();
+        } else {
+            warn!(
+                "tried to send a message to non-existent connection: {:?}",
+                to
+            );
+        }
+    }
+
+    pub fn broadcast(&mut self, message: Message) {
+        let inner = self.inner.lock().unwrap();
+        // ws-rs is a bit weird at broadcasting: you broadcast by obtaining a
+        // random connection, calling `broadcast` on it, and it will send a
+        // message to all connections made by the listener that this connection
+        // belongs to.
+        if let Some(connection) = inner.connections.values().next() {
+            let message = ws::Message::Text(String::from_utf8(message.data).unwrap());
+            connection.broadcast(message).log_if_err();
+        }
+    }
+}
+
+struct InnerServer {
+    next_connection_id: ConnectionId,
+    connections: HashMap<ConnectionId, ws::Sender>,
+}
+
+impl InnerServer {
+    fn generate_id(&mut self) -> ConnectionId {
+        let id = self.next_connection_id;
+        self.next_connection_id.0 += 1;
+        id
+    }
+}
+
+struct ConnectionHandler {
+    id: ConnectionId,
+    sender: Option<ws::Sender>,
+    events: Sender<Event>,
+    inner: Arc<Mutex<InnerServer>>,
+}
+
+impl ws::Handler for ConnectionHandler {
+    fn on_request(&mut self, req: &ws::Request) -> ws::Result<ws::Response> {
+        const INDEX_PATH: &str = "../client/target/index.html";
+        const JS_PATH: &str = "../client/target/bundle.js";
+        const SOURCE_MAP_PATH: &str = "../client/target/bundle.js.map";
+        const CSS_PATH: &str = "../client/target/style.css";
+
+        fn ok<P: AsRef<Path>>(path: P) -> io::Result<ws::Response> {
+            let mut file = fs::File::open(path)?;
+            let mut data = Vec::new();
+            file.read_to_end(&mut data)?;
+            Ok(ws::Response::new(200, "OK", data))
+        }
+
+        match req.resource() {
+            "/" => Ok(ok(INDEX_PATH)?),
+            "/bundle.js" => Ok(ok(JS_PATH)?),
+            "/bundle.js.map" => Ok(ok(SOURCE_MAP_PATH)?),
+            "/style.css" => Ok(ok(CSS_PATH)?),
+            "/ws" => {
+                self.events.send(Event::Connected { id: self.id }).unwrap();
+                let sender = self.sender
+                    .take()
+                    .expect("multiple websocket connection requests on single connection");
+                self.inner
+                    .lock()
+                    .unwrap()
+                    .connections
+                    .insert(self.id, sender);
+                ws::Response::from_request(req)
+            }
+            _ => Ok(ws::Response::new(
+                404,
+                "Not Found",
+                b"404 - Not Found".to_vec(),
+            )),
+        }
+    }
+
+    fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
```

Comment:
Is this guaranteed to react only to the messages made through a web socket and no other types (if there are any, I'm not sure)?

Reply:
`ws::Message` represents a websocket message - this handles websocket messages and nothing else.